### ISSUE TITLE
Make Modal wrapper configurable

### ISSUE THREAD

**Issue by mathieumg** [react-bootstrap/react-bootstrap#748]
Following discussion in https://github.com/mzabriskie/react-draggable/issues/56 it would be great if we could configure some more internals of the `Modal`, in this case configure what wraps the `Modal` content. We recently made changes to allow greater flexibility in that regard by making configurable some hardcoded `className`s, this would be the next step in that direction.

**Comment by mtscout6** [comment]
:+1: Bootstrap themselves don't want to support this, but if we can with minimal effort I see no reason why we shouldn't.

**Comment by cosminnicula** [comment]
:+1:

**Comment by cosminnicula** [comment]
You might find this useful https://github.com/jquense/react-bootstrap-modal

Notice how Modal is composed of Modal.Header, Modal.Body and Modal.Footer. In this way you have more control over the markup.

**Comment by taion** [comment]
:+1: I would love to be able to access just the overlay + hide on whatever logic and not use any of the standard Bootstrap modal styling.

**Comment by mtscout6** [comment]
@cosminnicula Thanks for sharing that modal by @jquense. I have to say that I REALLY like that approach, and it's in line with a lot of changes that I'd like to make to more of the components within this library. In fact I was doing something similar on the Dropdown re-write for the title so you could do:

```
<DropdownButton>
  <DropdownButtonTitle>Text</DropdownButtonTitle>
  <MenuItem>...</MenuItem>
</DropdownButtom>
```

I think that designing components in this fashion would also yield smaller simpler components to maintain and test.

@joemcbride and I just came up with a use case internally at work that would be a lot easier with @jquense's approach.

This would be a breaking change, but I think it would be for the better.

**Comment by jquense** [comment]
We've had great success with this sort of approach (with the modal and other components), its a bunch more markup but the added configurablility is well worth it. The convention of adding the "children" (`Modal.Header`) types as static properties reduces some of the added typing by still allowing a single import.

Please take anything ya'll can use from my code, its originally based of the modal here, so I'd be happy to contribute back.

**Comment by mtscout6** [comment]
@jquense I'd be open to seeing what a PR for this would look like. The hard part would be a sensible deprecation warning for the existing setup.

**Comment by jquense** [comment]
depends on what you want to break and what features are needed it seems.

It seems like the original ask here is to control all of the modal markup? That's easy enough for `.modal-body`, `.modal-header`, `modal-footer` but `modal-content` and `modal-dialog` become much tougher...Ideally you really do want to hide those away as they are structural markup needed to get the thing to work as opposed to a semantic piece of the modal

```js 

<Modal>
  <Modal.Dialog>
    <Modal.Content>
      <Modal.Body>...</Modal.Body>
    </Modal.Content>
  </Modal.Dialog>
</Modal>
```

feells preeettty ugly, I also think you need to control `Modal.Dialog` anyway otherwise you won't be able to animate it in. but in terms of a depreciation notice you can just check the children of `Modal` and duck type it, rendering the new way if it passes otherwise render the old way, for a version and then remove the old?

**Comment by mtscout6** [comment]
For deprecation that's the basic idea. A clean deprecation warning PR could come directly to master. We are also using pre-releases now for the next breaking change release to which the removal of said deprecations could also be pushed. That current breaking change branch is the `v0.24.0-rc` branch, future releases will be following a similar naming convention as outlined in the [Maintaining](https://github.com/react-bootstrap/react-bootstrap/blob/master/MAINTAINING.md#release-candidates) documentation.

As far as the various namespaced sub-components, I think you are more familiar with the Modal than I am. So, I'd trust you to make some good judgement calls there.

**Comment by jquense** [comment]
could I get a quick confirmation on what is needed to get this to work with react-draggable?  seems like the ability to specify the component with the `.modal-dialog` class?

**Comment by cosminnicula** [comment]
Saw that http://react-bootstrap.github.io/components.html#modals uses the new API: ```<Modal.Header/>```, ```<Modal.Title/>```, ```<Modal.Body/>``` and ```<Modal.Footer/>```

Can this issue be closed then?

**Comment by mathieumg** [comment]
Out of curiosity, did you try it with `react-draggable`? Did you manage to do what you wanted?

**Comment by cosminnicula** [comment]
Will try it and be back with feedback.

**Comment by jquense** [comment]
I don't actually think the crux of this issue is solved here. The changes still don't allow for custom component for the one that matters to react-draggable, the modal-dialog one. That is still wrapped up in the Modal component

**Comment by taion** [comment]
I think this is largely resolved by #1060. At most we might want to change the API to be consistent with #1096, but I think fuller customizability is better handled in react-overlays.
